`azurerm_storage_account` - Check `access_tier` for unsupported `storage_kind`

magodo opened the pull request:
`access_tier` is supported for `account_kind` of BlobStorage, StorageV2, and FileStorage. Previously, the behavior for unsupported `account_kind` is that the provider will silently allow users to set the `access_tier`, which causes the following plan/apply to show a diff (as the access tier is not really set).

This PR changes this behavior to error out for these supported `account_kind` cases.

## Test

```shell
terraform-provider-azurerm on  storage_account_access_tier_check via 🐹 v1.21.4
💤  TF_ACC=1 go test -v -timeout=20h ./internal/services/storage -run="TestAccStorageAccount_invalidAccountKindForAccessTier"
=== RUN   TestAccStorageAccount_invalidAccountKindForAccessTier
=== PAUSE TestAccStorageAccount_invalidAccountKindForAccessTier
=== CONT  TestAccStorageAccount_invalidAccountKindForAccessTier
--- PASS: TestAccStorageAccount_invalidAccountKindForAccessTier (3.92s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/storage       3.928s
```

Fix https://github.com/hashicorp/terraform-provider-azurerm/issues/24490
Fix https://github.com/hashicorp/terraform-provider-azurerm/issues/15210

katbyte reviewed (changes requested):
We have a new test failure on this pr:

```
------- Stdout: -------
=== RUN   TestAccStorageAccount_isHnsEnabled
=== PAUSE TestAccStorageAccount_isHnsEnabled
=== CONT  TestAccStorageAccount_isHnsEnabled
    testcase.go:113: Step 1/2 error: Error running apply: exit status 1
        
        Error: `is_hns_enabled` can only be used with account kinds `StorageV2`, `BlobStorage` and `BlockBlobStorage`
        
          with azurerm_storage_account.test,
          on terraform_plugin_test.tf line 26, in resource "azurerm_storage_account" "test":
          26: resource "azurerm_storage_account" "test" {
        
--- FAIL: TestAccStorageAccount_isHnsEnabled (111.88s)
FAIL
```

magodo commented:
@katbyte Thanks! I've updated and fixed it:

```shell
terraform-provider-azurerm on  storage_account_access_tier_check via 🐹 v1.21.6
💤  TF_ACC=1 go test -v -timeout=20h -run='TestAccStorageAccount_isHnsEnabled' ./internal/services/storage
=== RUN   TestAccStorageAccount_isHnsEnabled
=== PAUSE TestAccStorageAccount_isHnsEnabled
=== CONT  TestAccStorageAccount_isHnsEnabled
--- PASS: TestAccStorageAccount_isHnsEnabled (102.59s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/storage       102.609s
```

tombuildsstuff commented on internal/services/storage/storage_account_resource.go:
```diff
@@ -1246,6 +1247,20 @@ func resourceStorageAccount() *pluginsdk.Resource {
+
+				if d.Get("access_tier") != "" {
+					if accountKind := d.Get("account_kind").(string); !slices.Contains(
+						[]storage.Kind{
+							storage.KindBlobStorage,
+							storage.KindStorageV2,
+							storage.KindFileStorage,
+						},
```
this same slice of values is defined 3 times - can we pull this out into a constant/variable/function so that these don't fall out of sync?

magodo replied:
@tombuildsstuff Though the storage kinds that support access_tier (2 places) and hns (1 place) are the same, while they are for different checks, I've abstract them to two variables of list, though same value.

tombuildsstuff reviewed (changes requested):
Left a comment inline around maintainability, but this otherwise LGTM 👍

stephybun reviewed (commented):
@magodo 
Got some new test failures:
```
------- Stdout: -------
=== RUN   TestAccStorageAccount_isNFSv3Enabled
=== PAUSE TestAccStorageAccount_isNFSv3Enabled
=== CONT  TestAccStorageAccount_isNFSv3Enabled
    testcase.go:113: Step 1/2 error: Error running apply: exit status 1
        Error: `is_hns_enabled` can only be used with account of kinds: [BlobStorage StorageV2 FileStorage]
          with azurerm_storage_account.test,
          on terraform_plugin_test.tf line 41, in resource "azurerm_storage_account" "test":
          41: resource "azurerm_storage_account" "test" {
--- FAIL: TestAccStorageAccount_isNFSv3Enabled (160.62s)
FAIL
```
![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/25562180/660847d9-ba04-4e02-8a15-3bb41f8daed6)

magodo commented:
@stephybun Thank you Stephy! I've updated the code to pass these tests:


```shell
terraform-provider-azurerm on  storage_account_access_tier_check via 🐹 v1.21.6 took 8s
💤  TF_ACC=1 go test -v -timeout=20h -run='TestAccStorageAccount_isNFSv3Enabled|TestAccStorageAccount_hnsWithPremiumStorage|TestAccStorageAccount_invalidAccountKindForAccessTier' ./internal/services/storage
=== RUN   TestAccStorageAccount_isNFSv3Enabled
=== PAUSE TestAccStorageAccount_isNFSv3Enabled
=== RUN   TestAccStorageAccount_hnsWithPremiumStorage
=== PAUSE TestAccStorageAccount_hnsWithPremiumStorage
=== RUN   TestAccStorageAccount_invalidAccountKindForAccessTier
=== PAUSE TestAccStorageAccount_invalidAccountKindForAccessTier
=== CONT  TestAccStorageAccount_isNFSv3Enabled
=== CONT  TestAccStorageAccount_invalidAccountKindForAccessTier
=== CONT  TestAccStorageAccount_hnsWithPremiumStorage
--- PASS: TestAccStorageAccount_invalidAccountKindForAccessTier (1.49s)
--- PASS: TestAccStorageAccount_hnsWithPremiumStorage (150.41s)
--- PASS: TestAccStorageAccount_isNFSv3Enabled (198.14s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/storage       198.160s
```

stephybun reviewed (approved):
Thanks @magodo LGTM 👍

Poil commented:
Hi,

I still have the error on azurerm 3.93.0 provider.
Is this has been released ?

Regards
